fastcrypto-zkp/src/groth16/generic_api.rs

```diff
@@ -0,0 +1,210 @@
+/// bytes each, and serialized proof points.
+pub(crate) fn verify_groth16_in_bytes<
+    G1,
+    const G1_SIZE: usize,
+    const G2_SIZE: usize,
+    const GT_SIZE: usize,
+    const FR_SIZE: usize,
+>(
+    vk_gamma_abc_g1_bytes: &[u8],
+    alpha_g1_beta_g2_bytes: &[u8],
+    gamma_g2_neg_pc_bytes: &[u8],
+    delta_g2_neg_pc_bytes: &[u8],
+    proof_public_inputs_as_bytes: &[u8],
+    proof_points_as_bytes: &[u8],
+) -> Result<bool, FastCryptoError>
+where
+    G1: Pairing + ToFromByteArray<G1_SIZE> + for<'a> Deserialize<'a>,
+    <G1 as Pairing>::Other: ToFromByteArray<G2_SIZE> + for<'a> Deserialize<'a>,
+    <G1 as Pairing>::Output: GroupElement + ToFromByteArray<GT_SIZE>,
+    G1::ScalarType: ToFromByteArray<FR_SIZE>,
+{
+    let x = deserialize_vector::<FR_SIZE, G1::ScalarType>(proof_public_inputs_as_bytes)?;
+    let proof =
+        bcs::from_bytes(proof_points_as_bytes).map_err(|_| FastCryptoError::InvalidInput)?;
+    let prepared_vk = PreparedVerifyingKey::<G1>::deserialize_from_parts(
+        vk_gamma_abc_g1_bytes,
+        alpha_g1_beta_g2_bytes,
+        gamma_g2_neg_pc_bytes,
+        delta_g2_neg_pc_bytes,
+    )?;
+    Ok(prepared_vk.verify(&x, &proof).is_ok())
+}
+
+impl<G1: Pairing> VerifyingKey<G1> {
+    pub fn from_arkworks_format<const G1_SIZE: usize, const G2_SIZE: usize>(
+        bytes: &[u8],
+    ) -> FastCryptoResult<Self>
+    where
+        G1: ToFromByteArray<G1_SIZE>,
+        <G1 as Pairing>::Other: ToFromByteArray<G2_SIZE>,
+    {
+        // The verifying key consists of:
+        // - alpha: G1
+        // - beta: G2
+        // - gamma: G2
+        // - delta: G2
+        // - n: u64 lendian (size of gamma_abc)
+        // - gamma_abc: Vec<G1>
+
+        // We can't use bincode because there, the length of the vector is prefixed as a single byte
+        // and not a lendian u64.
+
+        if (bytes.len() - (G1_SIZE + 3 * G2_SIZE + 8)) % G1_SIZE != 0 {
+            return Err(FastCryptoError::InvalidInput);
+        }
+
+        let mut i = 0;
+
+        let alpha = G1::from_byte_array(
+            bytes[i..G1_SIZE]
+                .try_into()
+                .map_err(|_| FastCryptoError::InvalidInput)?,
+        )?;
+        i += G1_SIZE;
+
+        let beta = G1::Other::from_byte_array(
+            bytes[i..i + G2_SIZE]
+                .try_into()
+                .map_err(|_| FastCryptoError::InvalidInput)?,
+        )?;
+        i += G2_SIZE;
+
+        let gamma = G1::Other::from_byte_array(
+            bytes[i..i + G2_SIZE]
+                .try_into()
+                .map_err(|_| FastCryptoError::InvalidInput)?,
+        )?;
+        i += G2_SIZE;
+
+        let delta = G1::Other::from_byte_array(
+            bytes[i..i + G2_SIZE]
+                .try_into()
+                .map_err(|_| FastCryptoError::InvalidInput)?,
+        )?;
+        i += G2_SIZE;
+
+        let n = u64::from_le_bytes(
```

Comment:
should we verify that n>0?

Reply:
let's double check that this is backward compatible

Reply:
That's checked in `test_prepare_pvk_bytes` in api_tests.rs.